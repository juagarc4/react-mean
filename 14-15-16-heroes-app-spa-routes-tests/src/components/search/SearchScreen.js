import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom'
import queryString from 'query-string'

import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import { getHeroesByName } from '../../selectors/getHeroesByName'

export const SearchScreen = ({ history }) => {
  const location = useLocation()
  const { q = '' } = queryString.parse(location.search)
  const [formValues, handleInputChange] = useForm({ searchText: q })
  const { searchText } = formValues

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

  const handleSubmit = (e) => {
    e.preventDefault()
    history.push(`?q=${searchText}`)
  }
  return (
    <div className="row">
      <div className="col-md-5">
        <h4>Search Screen</h4>
        <hr />
        <form onSubmit={handleSubmit}>
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={handleInputChange}
            placeholder="Find a hero"
            type="text"
            value={searchText}
          />
          <div className="d-grid gap-2">
            <button
              type="submit"
              className="btn mt-2 btn-block btn-outline-primary"
            >
              Search
            </button>
          </div>
        </form>
      </div>
      <div className="col-md-7">
        <h4>Results</h4>
        <hr />

        {q === '' && <div className="alert alert-info">Search a hero</div>}
        {q !== '' && heroesFiltered.length === 0 && (
          <div className="alert alert-warning">
            There is no hero with the searched name <strong>"{q}"</strong>
          </div>
        )}

        {heroesFiltered.map((hero) => {
          return <HeroCard key={hero.id} {...hero} />
        })}
      </div>
    </div>
  )
}
