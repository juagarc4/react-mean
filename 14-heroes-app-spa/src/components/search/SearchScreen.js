import React from 'react'

import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'

export const SearchScreen = () => {
  const heroesFiltered = heroes
  const [formValues, handleInputChange] = useForm({ searchText: '' })
  const { searchText } = formValues
  const handleSubmit = (e) => {
    e.preventDefault()
    console.log(formValues)
  }
  return (
    <div className='row'>
      <div className='col-md-5'>
        <h4>Search Screen</h4>
        <hr />
        <form onSubmit={handleSubmit}>
          <input
            autoComplete='off'
            className='form-control'
            name='searchText'
            onChange={handleInputChange}
            placeholder='Find a hero'
            type='text'
            value={searchText}
          />
          <div class='d-grid gap-2'>
            <button type='submit' className='btn m-1 btn-block btn-outline-primary'>
              Search
            </button>
          </div>
        </form>
      </div>
      <div className='col-md-7'>
        <h4>Results</h4>
        <hr />
        {heroesFiltered.map((hero) => {
          return <HeroCard key={hero.id} {...hero} />
        })}
      </div>
    </div>
  )
}
