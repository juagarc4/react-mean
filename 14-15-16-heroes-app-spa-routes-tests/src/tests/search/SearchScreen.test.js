import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom'
import { SearchScreen } from 'components/search/SearchScreen'

describe('Tests in <SearchScreen/>', () => {
  test('Should show properly with default values', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero')
  })
  test('Should show Batman and the input with the value of the queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )
    expect(wrapper.find('input').prop('value')).toBe('batman')
    expect(wrapper).toMatchSnapshot()
  })
})
