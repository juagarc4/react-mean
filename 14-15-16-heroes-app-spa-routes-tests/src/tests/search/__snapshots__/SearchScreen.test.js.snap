// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen/> Should return an error if the hereo was not found. 1`] = `
<div
  className="row"
>
  <div
    className="col-md-5"
  >
    <h4>
      Search Screen
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find a hero"
        type="text"
        value="myhero"
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-md-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-warning"
    >
      There is no hero with the searched name 
      <strong>
        "
        myhero
        "
      </strong>
    </div>
  </div>
</div>
`;

exports[`Tests in <SearchScreen/> Should show Batman and the input with the value of the queryString 1`] = `
<div
  className="row"
>
  <div
    className="col-md-5"
  >
    <h4>
      Search Screen
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find a hero"
        type="text"
        value="batman"
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-md-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="card"
    >
      <div
        className="row g-0"
      >
        <div
          className="col-md-4"
        >
          <img
            alt="Batman"
            className="card-img rounded-0"
          />
        </div>
        <div
          className="col-md-6"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <h6
              className="card-subtitle mb-2 text-muted"
            >
              DC Comics
            </h6>
            <p
              className="card-text"
            >
              <small>
                Bruce Wayne
              </small>
            </p>
            <a
              className="align-text-bottom"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              More...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests in <SearchScreen/> Should show properly with default values 1`] = `
<div
  className="row"
>
  <div
    className="col-md-5"
  >
    <h4>
      Search Screen
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find a hero"
        type="text"
        value=""
      />
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-md-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-info"
    >
      Search a hero
    </div>
  </div>
</div>
`;
